libflux/src/libstd/build.rs

```diff
@@ -1,24 +1,62 @@
-use std::env;
-use std::fs::File;
-use std::io::Write;
-use std::path::PathBuf;
+use std::{env, fs, io, io::Write, path};
 
-fn prelude() -> Vec<u8> {
-    Vec::new()
+use flux::semantic::bootstrap;
+use flux::semantic::env::Environment;
+use flux::semantic::flatbuffers::types as fb;
+
+use flatbuffers;
+
+#[derive(Debug)]
+struct Error {
+    msg: String,
+}
+
+impl From<env::VarError> for Error {
+    fn from(err: env::VarError) -> Error {
+        Error {
+            msg: err.to_string(),
+        }
+    }
+}
+
+impl From<io::Error> for Error {
+    fn from(err: io::Error) -> Error {
+        Error {
+            msg: format!("{:?}", err),
+        }
+    }
 }
 
-fn stdlib() -> Vec<u8> {
-    Vec::new()
+impl From<bootstrap::Error> for Error {
+    fn from(err: bootstrap::Error) -> Error {
+        Error { msg: err.msg }
+    }
 }
 
-fn main() {
-    let dir = PathBuf::from(env::var("OUT_DIR").unwrap());
+fn serialize<'a, T, S, F>(ty: T, f: F, path: &path::Path) -> Result<(), Error>
+where
+    F: Fn(&mut flatbuffers::FlatBufferBuilder<'a>, T) -> flatbuffers::WIPOffset<S>,
+{
+    let mut builder = flatbuffers::FlatBufferBuilder::new();
+    let buf = fb::serialize(&mut builder, ty, f);
+    let mut file = fs::File::create(path)?;
+    file.write_all(&buf)?;
+    Ok(())
+}
+
+fn main() -> Result<(), Error> {
+    let dir = path::PathBuf::from(env::var("OUT_DIR")?);
+
+    let (pre, lib, fresher) = bootstrap::infer_stdlib()?;
+
+    let path = dir.join("prelude.data");
```

Comment:
Why does the prelude need to be serialized separately?  Isn't everything in there redundant with what's already in the standard lib?

Reply:
The prelude is composed of all identifiers that are defined in the `universe` and `influxdb` packages. These identifiers can be of any generic type. So the importer for the prelude associates identifier names with polytypes.

The importer for the standard library however associates import paths with polymorphic record (or row) types. A package can be thought of as an object. You use a member expression to access the identifiers defined in that package. That's why they're serialized separately.

Reply:
Ok, that makes sense to me.  Thanks.